pilot/pkg/model/xds_cache.go

```diff
@@ -164,17 +174,46 @@ func newLru(evictCallback simplelru.EvictCallback) simplelru.LRUCache {
 	return l
 }
 
-func (l *lruCache) evict(k interface{}, v interface{}) {
+// This is the callback passed to LRU, it will be called whenever a key is removed.
+func (l *lruCache) evict(k interface{}, _ interface{}) {
 	if features.EnableXDSCacheMetrics {
 		xdsCacheEvictions.Increment()
 	}
 
-	key := k.(string)
-	value := v.(cacheValue)
+	if l.recordEvicted {
+		l.evictedKeys.Insert(k.(string))
+	}
+}
 
-	// we don't need to acquire locks, since this function is called when we write to the store
-	l.clearConfigIndex(key, value.dependentConfigs)
-	l.clearTypesIndex(key, value.dependentTypes)
+// clearEvicted is run periodically to clear configIndex and typesIndex
+func (l *lruCache) clearEvicted() {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	// In order to save cpu, if the evicted keys is very small, we can just ignore
+	// TODO: make it configurable
+	if len(l.evictedKeys) < 100 {
+		return
+	}
+
+	for configKey, index := range l.configIndex {
```

Comment:
We have hundreds of customers using our mesh product(including multi clusters), and I confirmed their scalability could never exceed 60_000 size.



So why 99.9% users have to pay additional memory cost for which could never be used?


BTW, istio performance test in 1.15 show much more memory usage than before seems very bad.

Reply:
If they do not need, why do not they reduce the xds_cache_size?

> So why 99.9% users have to pay additional memory cost for which could never be used?

I doubt this part. The cluster where I have shared the data is not one of our biggest clusters. I would say it is a medium sized clusters. But the routes have many delegates - that is why DependentConfigs size has grown and delegate keys are most occupied. I doubt this PR will fix that because we any have all the DependentConfigs in xds cache struct.

Reply:
Whatever, the pr is trying to remove cache values storing dependent configs, benefit to any users.

Reply:
We got a bit off topic here but my original comment still stands that we need performance numbers on this change before submitting. Right now we are just guessing if it's better (as far as I can tell, unless you have data not shared?)

Reply:
nope, this is not fixing the high memory cost issue. It is reverting a change.